src/ResourceManager/Resources/Commands.Resources/ActiveDirectory/NewAzureADServicePrincipalCommand.cs

```diff
@@ -133,53 +216,180 @@ public override void ExecuteCmdlet()
+            };
+
+            if (!this.IsParameterBound(c => c.ApplicationId))
+            {
+                CreatePSApplicationParameters appParameters = new CreatePSApplicationParameters
+                {
+                    DisplayName = DisplayName,
+                    IdentifierUris = new[] { identifierUri },
+                    HomePage = identifierUri,
+                    PasswordCredentials = new PSADPasswordCredential[]
+                    {
+                        passwordCredential
+                    }
+                };
+
+                if (ShouldProcess(target: appParameters.DisplayName, action: string.Format("Adding a new application for with display name '{0}'", appParameters.DisplayName)))
+                {
+                    var application = ActiveDirectoryClient.CreateApplication(appParameters);
+                    ApplicationId = application.ApplicationId;
+                    WriteVerbose(string.Format("No application id provided - created new AD application with application id '{0}'", ApplicationId));
+                }
+            }
+
+            CreatePSServicePrincipalParameters createParameters = new CreatePSServicePrincipalParameters
+            {
+                ApplicationId = ApplicationId,
+                AccountEnabled = true,
+                PasswordCredentials = new PSADPasswordCredential[]
+                {
+                    passwordCredential
+                }
+            };
+
+            if (ShouldProcess(target: createParameters.ApplicationId.ToString(), action: string.Format("Adding a new service principal to be associated with an application having AppId '{0}'", createParameters.ApplicationId)))
+            {
+                var servicePrincipal = ActiveDirectoryClient.CreateServicePrincipal(createParameters);
+                WriteObject(servicePrincipal);
+                if (this.IsParameterBound(c => c.SkipAssignment))
+                {
+                    WriteVerbose("Skipping role assignment for the service principal.");
+                    return;
+                }
+
+                FilterRoleAssignmentsOptions parameters = new FilterRoleAssignmentsOptions()
+                {
+                    Scope = this.Scope,
+                    RoleDefinitionName = this.Role,
+                    ADObjectFilter = new ADObjectFilterOptions
+                    {
+                        SPN = servicePrincipal.ApplicationId.ToString(),
+                        Id = servicePrincipal.Id.ToString()
+                    },
+                    ResourceIdentifier = new ResourceIdentifier()
+                    {
+                        Subscription = subscriptionId
+                    },
+                    CanDelegate = false
+                };
+
+                for (var i = 0; i < 6; i++)
+                {
+                    try
+                    {
+                        Thread.Sleep(5000);
```

Comment:
We should use our own TestMockSupport.Delay method so this doesn't take forever in playback